### PR TITLE
Aem download manger

### PR DESCRIPTION
**Pull request by gyasistory**
This is all of the functionality for pulling Articles from a Manifest and saving it to the database.

### CODE REVIEW

**Comment by gyasistory** [comment]
Found a possible issue with Android 15.  The Uri conversion doesn't seem to work.

File: app/build.gradle [hunk 1]
```diff
@@ -171,11 +171,15 @@ dependencies {
 
 // make crashlytics upload task dependent on the assemble task for all build variants
 afterEvaluate {
+
+//    crashlyticsUploadDistributionProductionDebug.dependsOn assembleProductionDebug
+
```
Comment (frett) [inline app/build.gradle]: unnecessary change, can be removed

File: app/build.gradle [hunk 2]
```diff
@@ -171,11 +171,15 @@ dependencies {
...
     android.applicationVariants.all { variant ->
         def assembleTask = project.tasks.getByName("assemble${variant.name.capitalize()}")
         def crashlyticsTask = project.tasks.getByName("crashlyticsUploadDistribution${variant.name.capitalize()}")
         crashlyticsTask.dependsOn assembleTask
     }
+
```
Comment (frett) [inline app/build.gradle]: unnecessary change

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java [hunk 3]
```diff
@@ -95,20 +100,70 @@ public void verifyGetAllArticles() {
...
     @Test
     public void verifyManifestHasArticles() {
         assertTrue("Manifest 0 should have articles", mAssociationDao.getTestableArticlesByManifestID("0").size() > 0);
         assertTrue("Manifest 1 should have articles", mAssociationDao.getTestableArticlesByManifestID("1").size() > 0);
         assertTrue("Manifest 2 should have articles", mAssociationDao.getTestableArticlesByManifestID("2").size() > 0);
     }
+
+
+    @Test
+    public void verifyArticleParseLogic() {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java]: This should be in a separate `ArticleParserTest` class.

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java [hunk 4]
```diff
@@ -95,20 +100,70 @@ public void verifyGetAllArticles() {
...
+
+        try {
+            Uri url = Uri.parse("https://stage.cru.org/content/experience-fragments/questions_about_god/english.999.json");
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java]: by depending on a web resource this test is very brittle. You should embed a raw json file and load that to perform your parsing tests. That way if the web content ever changes it doesn't break your test

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java [hunk 5]
```diff
@@ -95,20 +100,70 @@ public void verifyGetAllArticles() {
...
+        }
+
+    }
+
+    @Test
+    public void verifyAttachmentsAreSaved() {
+        for (Article article : mSavedArticles) {
+            for (Attachment attachment : mAttachmentDao.getTestableAttachmentsByArticle(article.mkey)) {
+                try {
+                    AEMDownloadManger
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java]: you are testing AEMDownloadManager functionality, so this should be in an `AEMDownloadManagerTest` test class instead of in the `ArticleRoomDatabaseTest` class

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    public static final String ATTACHMENT_LIST_KEY = "attachment_list_key";
+
+    private JSONObject articleJSON;
+
+    private List<Attachment> attachmentList = new ArrayList<>();
+
+    private List<Article> articleList = new ArrayList<>();
+
+    private String createdTag = "jcr:created";
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: these should be `private static final` to make them constants.

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    private String contentTag = "jcr:content";
+    private String lastModifiedTag = "cq:lastModified";
+    private String uuidTag = "jcr:uuid";
+    private String titleTag = "jcr:title";
+    private String rootTag = "root";
+    private String fileTag = "fileReference";
+    private String baseUrl = "https://stage.cru.org";
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: this shouldn't be here, where are you using it?

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: you should be relying on the AEM import url for whatever functionality this is providing

File: library/db/build.gradle [hunk 8]
```diff
@@ -8,4 +8,5 @@ dependencies {
     api "org.ccci.gto.android:gto-support-db-stream:${deps.gtoSupport}"
 
     implementation "com.crashlytics.sdk.android:crashlytics:${deps.crashlytics}"
+
```
Comment (frett) [inline library/db/build.gradle]: extraneous whitespace, shouldn't have been committed

File: ui/articles-aem-renderer/build.gradle [hunk 9]
```diff
@@ -5,14 +5,24 @@ android {
...
     annotationProcessor "android.arch.lifecycle:compiler:${deps.androidLifecycle}"
     annotationProcessor "android.arch.persistence.room:compiler:${deps.roomVersion}"
+
+    // for convert Uri to Json
+    implementation "org.apache.directory.studio:org.apache.commons.io:${deps.ioCommons}"
+
+    implementation "com.jakewharton.timber:timber:${deps.timber}"
+
 }
+
```
Comment (frett) [inline ui/articles-aem-renderer/build.gradle]: extraneous whitespace

File: ui/articles-aem-renderer/build.gradle [hunk 10]
```diff
@@ -5,14 +5,24 @@ android {
...
+    implementation project(':library:xml-model')
 
-    implementation "android.arch.lifecycle:extensions:${deps.androidLifecycle}"
-    implementation "android.arch.persistence.room:runtime:${deps.roomVersion}"
-    implementation "com.android.support:appcompat-v7:${deps.androidSupport}"
-
+    // For Testing
     androidTestImplementation "com.android.support.test:runner:${deps.androidTestingSupport}"
     androidTestImplementation "android.arch.persistence.room:testing:${deps.roomVersion}"
+    testImplementation "android.arch.core:core-testing:${deps.androidLifecycle}"
```
Comment (frett) [inline ui/articles-aem-renderer/build.gradle]: this dependency does nothing since it is targeting JVM unit tests and not the android tests.

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java [hunk 11]
```diff
@@ -95,20 +100,70 @@ public void verifyGetAllArticles() {
...
+            for (Attachment attachment : mAttachmentDao.getTestableAttachmentsByArticle(article.mkey)) {
+                assertFalse(attachment.mAttachmentFilePath != null);
+            }
+        }
+    }
 }
+
+
+
+
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleRoomDatabaseTest.java]: extraneous trailing whitespace

**Comment by frett** [comment]
we should use OkHttp3 for requests, that should handle the SSL protocol correctly

**Comment by frett** [comment]
plus, we are already utilizing OkHttp3 for all other web requests (via Retrofit)

**Comment by gyasistory** [comment]
OK I will try using that.

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java (hunk 7 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: I am just using them to parse the json data.  I wanted to make them member variable so I can easily change them if needed.

File: app/build.gradle [hunk 12]
```diff
@@ -171,6 +171,9 @@ dependencies {
 
 // make crashlytics upload task dependent on the assemble task for all build variants
 afterEvaluate {
+
+    crashlyticsUploadDistributionProductionDebug.dependsOn assembleProductionDebug
+
```
Comment (frett) [inline app/build.gradle]: this code should be removed, not uncommented. it was replaced by the `android.applicationVariants.all` code right below it

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/ArticleParseTest.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+            OkHttpClient client = new OkHttpClient();
+
+            Request request = new Request.Builder()
+                    .url(url.toString())
+                    .build();
+
+            Response response = client.newCall(request).execute();
+
+            String result = response.body().string();
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/ArticleParseTest.java]: for tests you shouldn't be relying on fetching content from the web, it makes a test very brittle. you should include the json as a file and parse that. This makes the test self-contained and not dependent on an external system. see:
https://github.com/CruGlobal/godtools-android/blob/develop/library/xml-model/src/androidTest/java/org/cru/godtools/xml/model/ManifestIT.java#L18-L23
https://github.com/CruGlobal/godtools-android/blob/develop/library/xml-model/src/androidTest/assets/tests/manifest-empty.xml

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/ArticleParseTest.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+            HashMap<String, Object> results = new ArticleParser(jsonObject).execute();
+
+            assertTrue(results.containsKey(ArticleParser.ATTACHMENT_LIST_KEY));
+            assertTrue(results.containsKey(ArticleParser.ARTICLE_LIST_KEY));
+
+            List<Article> resultsArticles = (List<Article>) results.get(ArticleParser.ARTICLE_LIST_KEY);
+            assertTrue(resultsArticles.size() > 0);
+            List<Attachment> resultAttachments = (List<Attachment>) results.get(ArticleParser.ATTACHMENT_LIST_KEY);
+            assertTrue(resultAttachments.size() > 0);
+        } catch (JSONException | IOException | IllegalStateException e) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/ArticleParseTest.java]: what you can do instead of catching the exceptions, is just declare the test with `throws Exception`, then you won't have to catch it only to propagate it as a `fail`. within tests you shouldn't worry about graceful error handling at all, tests should throw exception when something unexpected happens.

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    /**
+     * This execute function allows you to change the initial jsonObject and run the execution.
+     *
+     * use <code>ARTICLE_LIST_KEY</code> and <code>ATTACHMENT_LIST_KEY</code> to access the Collections
+     * out of the HashMaps
+     *
+     * @param jsonObject = the object to be parsed.
+     * @return = return a list of <code>Article</code> and <code>Attachments</code> in HashMap
+     */
+    public HashMap<String, Object> execute(JSONObject jsonObject) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: I don't think you ever use this version of `execute`

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: from an overall class design perspective, this class has potential for data leakage if the same Parser object is used multiple times due to how the instance variables are used. There can be the intention that it will never be used in that way, but a developer a couple years down the road may not understand that constraint and mis-use the class.

Using instance variables is fine for tracking state, but I would make it so that `articleJSON` is final and only set in the constructor, and ensure you only parse the json one time when creating the object.

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+     * @param evaluatingObject = the Json Object to be evaluated.
+     */
+    private void jsonObjectHandler(JSONObject evaluatingObject) {
+        // Create loop through Keys
+        Iterator<String> keys = evaluatingObject.keys();
+        while (keys.hasNext()) {
+            String nextKey = keys.next();
+            if (isArticleOrCategory(nextKey)) {
+                try {
+                    JSONObject returnedObject = evaluatingObject.getJSONObject(nextKey);
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: I probably would have used this structure instead:
```java
JSONObject returnedObject = evaluatingObject.optJSONObject(nextKey);
if(returnedObject != null) {
  ...
}
```

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: then you wouldn't have needed to use a `catch`

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+                    if (isObjectCategory(returnedObject)) {
+                        jsonObjectHandler(returnedObject);
+                    } else {
+                        // Get inner Article Object
+                        parseArticleObject(returnedObject);
+                    }
+                } catch (Exception e) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: catching all `Exceptions` like this can hide bugs

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+        // Create loop through Keys
+        Iterator<String> keys = evaluatingObject.keys();
+        while (keys.hasNext()) {
+            String nextKey = keys.next();
+            if (isArticleOrCategory(nextKey)) {
+                try {
+                    JSONObject returnedObject = evaluatingObject.getJSONObject(nextKey);
+                    if (isObjectCategory(returnedObject)) {
+                        jsonObjectHandler(returnedObject);
+                    } else {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: I don't know for sure, but I would assume there might be other types of "objects" that may be present in the json. it's probably safer to use something like:
```java
switch(returnedObject.optString("jcr:primaryType", "")) {
  case "sling:OrderedFolder":
    jsonObjectHandler(returnedObject);
    break;
  case "{{page_type}}":
    parseArticleObject(returnedObject);
    break;
}
```

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: also, `jcr:primaryType`, `sling:OrderedFolder`, and whatever the page type is should be static private constants defined at the top of this class

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+                }
+            }
+        }
+    }
+
+    /**
+     * This method parses an Article Json Object into the Database. On completion it will add
+     * articles to <code>articleList</code>
+     *
+     * @param articleObject = article JsonObject
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: the actual javadoc format for parameters doesn't use an `=`, it should just be:
```
* @param articleObject article JsonObject
```
example: https://www.oracle.com/technetwork/articles/java/index-137868.html#format

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+                    .getJSONObject(CONTENT_TAG).getString(LAST_MODIFIED_TAG));
+        }
+        retrievedArticle.mkey = articleObject.getJSONObject(CONTENT_TAG).getString(UUID_TAG);
+        retrievedArticle.mTitle = articleTagObject.getJSONObject(CONTENT_TAG)
+                .getString(TITLE_TAG);
+
+        JSONObject articleRootObject = articleTagObject.getJSONObject(CONTENT_TAG)
+                .getJSONObject(ROOT_TAG);
+
+        retrievedArticle.mContent = articleRootObject.getJSONObject("text").getString("text");
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: this line can be removed, the actual content will be downloaded elsewhere

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+     *
+     * @param articleRootObject = the root json Object of Article
+     * @param articleKey = the uuid of the article
+     */
+    private void getAttachmentsFromRootObject(JSONObject articleRootObject, String articleKey) {
+        // Iterate through keys
+        Iterator<String> keys = articleRootObject.keys();
+        while (keys.hasNext()) {
+            String nextKey = keys.next();
+            if (nextKey.contains("image")) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: you shouldn't be looking for images based on the key name, you should look based on this attribute:
```
"sling:resourceType": "wcm/foundation/components/image"
```

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java (hunk 15 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: Would it make more since to make execute a static method that takes in a local JsonObject, and use the local object to parse?

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/AEMDownloadManger.java [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+            String[] urlSplit = attachment.mAttachmentUrl.split("/");
+            String filename = urlSplit[urlSplit.length - 1];
+            File articleFile = new File(context.getFilesDir(), "articles");
+            if (!articleFile.exists()) {
+                articleFile.mkdir();
+            }
+            articleFile = new File(articleFile, filename);
+            FileOutputStream outputStream = new FileOutputStream(articleFile);
+            URL url = new URL(attachment.mAttachmentUrl);
+            outputStream.write(IOUtils.toByteArray(url));
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/AEMDownloadManger.java]: this should be using `OkHttp`

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+ */
+public class ArticleParser {
+
+    public static final String ARTICLE_LIST_KEY = "article_list_key";
+
+    public static final String ATTACHMENT_LIST_KEY = "attachment_list_key";
+
+    private static List<Attachment> attachmentList = new ArrayList<>();
+
+    private static List<Article> articleList = new ArrayList<>();
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/service/support/ArticleParser.java]: these 2 lists are still going to cause problems. They aren't cleared when parsing a second article json object.

**Comment by frett** [comment]
you will need to resolve merge conflicts with the develop branch

**Comment by frett** [comment]
@gyasistory you can look at the individual commits I made to get an idea of my thought process.
